File: objective-c-sample/JSQMessagesViewController_SendBird/ViewController.m
//
//  ViewController.m
//  JSQMessagesViewController_SendBird
//
//  Created by Jed Gyeong on 4/23/16.
//  Copyright © 2016 SENDBIRD. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UITextField *userIdTextField;
@property (weak, nonatomic) IBOutlet UITextField *nicknameTextField;
@property (weak, nonatomic) IBOutlet UIButton *connectButton;
@property (weak, nonatomic) IBOutlet UIButton *openChannelButton;
@property (weak, nonatomic) IBOutlet UIButton *groupChannelButton;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicatorView;
@property (weak, nonatomic) IBOutlet UILabel *versionLabel;

@property (atomic) BOOL connected;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    NSString *userId = [[NSUserDefaults standardUserDefaults] objectForKey:@"sendbird_user_id"];
    NSString *userName = [[NSUserDefaults standardUserDefaults] objectForKey:@"sendbird_user_name"];
    
    [self.userIdTextField setText:userId];
    [self.nicknameTextField setText:userName];
    
    self.connected = NO;
    
    [self.activityIndicatorView setHidden:YES];
    [self.openChannelButton setEnabled:NO];
    [self.groupChannelButton setEnabled:NO];
    
    NSString *path = [[NSBundle mainBundle] pathForResource:@"Info" ofType:@"plist"];
    if (path != nil) {
        NSDictionary *infoDict = [NSDictionary dictionaryWithContentsOfFile:path];
        NSString *sampleUIVersion = infoDict[@"CFBundleShortVersionString"];
        NSString *version = [NSString stringWithFormat:@"SDK v%@\nSample UI for Objective-C v%@", [SBDMain getSDKVersion], sampleUIVersion];
        self.versionLabel.text = version;
    }
}

- (IBAction)clickConnectButton:(id)sender {
    if (self.userIdTextField.text.length == 0) {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:@"User ID is required." preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *closeAction = [UIAlertAction actionWithTitle:@"Close" style:UIAlertActionStyleCancel handler:nil];
        [alert addAction:closeAction];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self presentViewController:alert animated:YES completion:nil];
        });
        
        return;
    }
    
    if (self.nicknameTextField.text.length == 0) {
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:@"Nickname is required." preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *closeAction = [UIAlertAction actionWithTitle:@"Close" style:UIAlertActionStyleCancel handler:nil];
        [alert addAction:closeAction];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self presentViewController:alert animated:YES completion:nil];
        });
        
        return;
    }
    
    if (self.connected) {
        [SBDMain disconnectWithCompletionHandler:^{
            self.connected = NO;
            dispatch_async(dispatch_get_main_queue(), ^{
                self.openChannelButton.enabled = NO;
                self.groupChannelButton.enabled = NO;
                self.userIdTextField.enabled = YES;
                [self.connectButton setTitle:@"Connect" forState:UIControlStateNormal];
            });
        }];
    }
    else {
        dispatch_async(dispatch_get_main_queue(), ^{
            self.activityIndicatorView.hidden = NO;
            [self.activityIndicatorView startAnimating];
        });
        [self.userIdTextField setEnabled:NO];
        [SBDMain connectWithUserId:[self.userIdTextField text] completionHandler:^(SBDUser * _Nullable user, SBDError * _Nullable error) {
            if (error == nil) {
                [SBDMain updateCurrentUserInfoWithNickname:[self.nicknameTextField text] profileUrl:nil completionHandler:^(SBDError * _Nullable error) {
                    if (error != nil) {
                        UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:[NSString stringWithFormat:@"%ld: %@", error.code, error.domain] preferredStyle:UIAlertControllerStyleAlert];
                        UIAlertAction *closeAction = [UIAlertAction actionWithTitle:@"Close" style:UIAlertActionStyleCancel handler:nil];
                        [alert addAction:closeAction];
                        dispatch_async(dispatch_get_main_queue(), ^{
                            [self presentViewController:alert animated:YES completion:nil];
                            [self.activityIndicatorView setHidden:YES];
                            [self.activityIndicatorView stopAnimating];
                        });
                        
                        return;
                    }
                    
                    [[NSUserDefaults standardUserDefaults] setObject:[SBDMain getCurrentUser].userId forKey:@"sendbird_user_id"];
                    [[NSUserDefaults standardUserDefaults] setObject:[SBDMain getCurrentUser].nickname forKey:@"sendbird_user_name"];
                    
                    dispatch_async(dispatch_get_main_queue(), ^{
                        self.connected = YES;
                        [self.openChannelButton setEnabled:YES];
                        [self.groupChannelButton setEnabled:YES];
                        [self.connectButton setTitle:@"Disconnect" forState:UIControlStateNormal];
                        
                        [self.activityIndicatorView stopAnimating];
                        [self.activityIndicatorView setHidden:YES];
                    });
                }];
            }
            else {
                UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:[NSString stringWithFormat:@"%ld: %@", error.code, error.domain] preferredStyle:UIAlertControllerStyleAlert];
                UIAlertAction *closeAction = [UIAlertAction actionWithTitle:@"Close" style:UIAlertActionStyleCancel handler:nil];
                [alert addAction:closeAction];
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self presentViewController:alert animated:YES completion:nil];
                    
                    [self.userIdTextField setEnabled:YES];
                    
                    [self.activityIndicatorView setHidden:YES];
                    [self.activityIndicatorView stopAnimating];
                });
            }
        }];
    }
}

- (IBAction)clickOpenChannelButon:(id)sender {
    OpenChannelListViewController *vc = [[OpenChannelListViewController alloc] init];
    [self.navigationController pushViewController:vc animated:YES];
}

- (IBAction)clickGroupChannelButton:(id)sender {
    GroupChannelListViewController *vc = [[GroupChannelListViewController alloc] init];
    [vc setUserID:[SBDMain getCurrentUser].userId userName:[SBDMain getCurrentUser].nickname];
    [self.navigationController pushViewController:vc animated:YES];
}

@end

File: swift-sample/SampleUI/ViewController.swift
//
//  ViewController.swift
//  SampleUI
//
//  Created by Jed Kyung on 8/19/16.
//  Copyright © 2016 SendBird. All rights reserved.
//

import UIKit
import SendBirdSDK

class ViewController: UIViewController {
    @IBOutlet weak var userIdTextField: UITextField!
    @IBOutlet weak var nicknameTextField: UITextField!
    @IBOutlet weak var connectButton: UIButton!
    @IBOutlet weak var openChannelButton: UIButton!
    @IBOutlet weak var groupChannelButton: UIButton!
    @IBOutlet weak var activityIndicatorView: UIActivityIndicatorView!
    @IBOutlet weak var versionLabel: UILabel!
    
    var connected: Bool
    
    required init?(coder aDecoder: NSCoder) {
        self.connected = false
        
        super.init(coder: aDecoder)
    }
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        self.connected = false
        
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let userId = UserDefaults.standard.object(forKey: "sendbird_user_id") as? String
        let nickname = UserDefaults.standard.object(forKey: "sendbird_nickname") as? String
        
        self.userIdTextField.text = userId
        self.nicknameTextField.text = nickname
        
        self.connected = false
        
        self.activityIndicatorView.isHidden = true
        self.openChannelButton.isEnabled = false
        self.groupChannelButton.isEnabled = false
        
        let path = Bundle.main.path(forResource: "Info", ofType: "plist")
        if path != nil {
            let infoDict = NSDictionary(contentsOfFile: path!)
            let sampleUIVersion = infoDict!["CFBundleShortVersionString"] as! String
            let version = String(format: "SDK v%@\nSample UI v%@ for Swift 2.2 ", SBDMain.getSDKVersion(), sampleUIVersion)
            self.versionLabel.text = version
        }
    }
    
    @IBAction func clickConnectButton(_ sender: AnyObject) {
        if self.userIdTextField.text?.characters.count == 0 {
            let alert = UIAlertController(title: "Error", message: "User ID is required.", preferredStyle: UIAlertControllerStyle.alert)
            let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
            alert.addAction(closeAction)
            DispatchQueue.main.async(execute: {
                self.present(alert, animated: true, completion: nil)
            })
            
            return;
        }
        
        if self.nicknameTextField.text?.characters.count == 0 {
            let alert = UIAlertController(title: "Error", message: "Nickname is required.", preferredStyle: UIAlertControllerStyle.alert)
            let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
            alert.addAction(closeAction)
            DispatchQueue.main.async(execute: {
                self.present(alert, animated: true, completion: nil)
            })
            
            return;
        }
        
        if self.connected {
            SBDMain.disconnect(completionHandler: {
                self.connected = false
                DispatchQueue.main.async(execute: {
                    self.openChannelButton.isEnabled = false
                    self.groupChannelButton.isEnabled = false
                    self.userIdTextField.isEnabled = true
                    self.connectButton.setTitle("Connect", for: UIControlState())
                })
            })
        }
        else {
            DispatchQueue.main.async(execute: {
                self.activityIndicatorView.isHidden = false
                self.activityIndicatorView.startAnimating()
            })
            self.userIdTextField.isEnabled = false
            SBDMain.connect(withUserId: self.userIdTextField.text!, completionHandler: { (user, error) in
                if error == nil {
                    SBDMain.updateCurrentUserInfo(withNickname: self.nicknameTextField.text, profileUrl: nil, completionHandler: { (error) in
                        if error != nil {
                            let alert = UIAlertController(title: "Error", message: String(format: "%lld: %@", error!.code, (error?.domain)!), preferredStyle: UIAlertControllerStyle.alert)
                            let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
                            alert.addAction(closeAction)
                            DispatchQueue.main.async(execute: {
                                self.present(alert, animated: true, completion: nil)
                                self.activityIndicatorView.isHidden = true
                                self.activityIndicatorView.stopAnimating()
                            })
                            
                            return
                        }
                        
                        UserDefaults.standard.set(SBDMain.getCurrentUser()?.userId, forKey: "sendbird_user_id")
                        UserDefaults.standard.set(SBDMain.getCurrentUser()?.nickname, forKey: "sendbird_nickname")
                        
                        DispatchQueue.main.async(execute: { 
                            self.connected = true
                            self.openChannelButton.isEnabled = true
                            self.groupChannelButton.isEnabled = true
                            self.connectButton.setTitle("Disconnect", for: UIControlState())
                            
                            self.activityIndicatorView.isHidden = true
                            self.activityIndicatorView.stopAnimating()
                        })
                    })
                }
                else {
                    let alert = UIAlertController(title: "Error", message: String(format: "%lld-%@", error!.code, (error?.domain)!), preferredStyle: UIAlertControllerStyle.alert)
                    let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
                    alert.addAction(closeAction)
                    DispatchQueue.main.async(execute: {
                        self.present(alert, animated: true, completion: nil)
                        
                        self.userIdTextField.isEnabled = true
                        
                        self.activityIndicatorView.isHidden = true
                        self.activityIndicatorView.stopAnimating()
                    })
                }
            })
        }
    }
    
    @IBAction func clickOpenChannelButon(_ sender: AnyObject) {
        let vc = OpenChannelListViewController()
        self.navigationController?.pushViewController(vc, animated: true)
    }
    
    @IBAction func clickGroupChannelButton(_ sender: AnyObject) {
        let vc = GroupChannelListViewController()
        vc.setUser((SBDMain.getCurrentUser()?.userId)!, aUserName: (SBDMain.getCurrentUser()?.nickname)!)
        self.navigationController?.pushViewController(vc, animated: true)
    }
}

